docs/src/proposals/off-chain-message-signing.md

```diff
@@ -0,0 +1,78 @@
+# Off-chain message signing
+## Motivation
+
+There is ecosystem demand for a method of signing non-transaction messages with
+a Solana wallet. Typically this is some kind of "proof of wallet ownership" for
+entry into a whitelisted system.
+
+Some inspiration can be gleaned from relevant portions of Ethereum's
+[EIP-721](https://eips.ethereum.org/EIPS/eip-712)
+
+## Considerations
+
+* Security
+  * Off-chain message signatures should not be valid transaction message signatures
+* Future-proofing
+  * Versioning
+  * Co-exist with versioned transaction messages and extended length transactions
+* Compatibility
+  * Hardware wallet signing
+  * Localization
+
+## Payload format
+
+| 0x00  ..= 0x0f | 0x10           | 0x11           | 0x12  ..= 0x13 | 0x14 .. |
+| :------------: | :------------: | :------------: | :------------: | :-----: |
+| Signing Domain | Header Version | Message Format | Message Length | Message |
+
+The payload consists of a 16-byte signing domain specifier, 4-byte header and
+n-byte message.
+
+### Signing Domain Specifier
+The signing domain specifier is a prefix byte string used to give similar structure
+to all off-chain message signatures. We assign its value to be:
+```
+b"\xffsolana offchain"
+```
+The first byte, `\xff`, was chosen for the following reasons
+1. It corresponds to a value that is illegal as the first byte in a transaction
+`MessageHeader`.
+1. It avoids unintentional misuse in languages with C-like, null-terminated strings.
+
+The remaining bytes, `b"solana offchain"`, were chosen to be descriptive and
+reasonably long, but are otherwise arbitrary.
+
+### Header
+The header consists of an 8-bit version. Version `0` headers have two additional
+fields; an 8-bit message format identifier and a 16-bit message length.
+
+#### Message Formats
+Version `0` headers specify three message formats allowing for trade-offs between
+compatibility and composition of messages.
+
+| ID  | Encoding            | Maximum Length | Harware Wallet Support |
+| :-: | :-----------------: | :------------: | :--------------------: |
+|  0  | Restricted ASCII \* | 1212           | Yes                    |
+|  1  | UTF-8               | 1212           | Blind sign only        |
+|  2  | UTF-8               | 65515          | No                     |
+
+\* Those characters for which [`isprint(3)`](https://linux.die.net/man/3/isprint)
+returns true.  That is, `0x20..=0x7f`.
+
+Both the message encoding and maximum message length **MUST** be enforced by both
+signer and verifier.
+
+Formats `0` and `1` are motivated by hardware wallet support where both RAM
+to store the payload and font character support are limited.
+
```

Comment:
MAYBE: should there be a section on signing of off-chain messages here?

1. Indicate that Solana messages should be signed with the same Ed25519 signature algorithm as Solana transactions
2. Note that the entire payload (header + message) should be included in the signature calculation
3. Standardize the signature(s) should be appended to the payload

Reply:
> 3. Standardize the signature(s) should be appended to the payload

good call. we should specify a container akin to `Transaction`.  may need to support multisig.  i'll noodle on it while i'm ooo in case there's anything else that it should support